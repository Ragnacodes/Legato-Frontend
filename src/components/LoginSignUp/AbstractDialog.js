import React from 'react';
import CloseRoundedIcon from "@material-ui/icons/CloseRounded";
import Typography from '@material-ui/core/Typography';
import Dialog from '@material-ui/core/Dialog';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';

const AbstractDialog = ({ open, onClick, title, description, form }) => {
    return (
        <Dialog
            disableBackdropClick
            className="dialog"
            open={open}
            onClose={() => onClick(false)}
            aria-labelledby="form-dialog-title"
        >
            <DialogTitle disableTypography={true} id="form-dialog-title">
                <Typography variant="h5">{title}</Typography>
            </DialogTitle>

            <DialogContent>
                <CloseRoundedIcon
                    className="close-icon"
                    onClick={() => onClick(false)}
                />
                <DialogContentText>{description}</DialogContentText>
                {form}
            </DialogContent>
        </Dialog>
    );
};

export default AbstractDialog;