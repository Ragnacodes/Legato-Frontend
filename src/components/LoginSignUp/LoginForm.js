import React from "react";
import { connect } from "react-redux";
import axios from "axios";

import * as actions from "../../actions/login";
import { login } from "../../actions/auth";

import { TextField, Button } from "@material-ui/core";

import { successNotification, errorNotification } from "../Notification";

export function LoginForm({
  errors,
  info,
  updateInfo,
  validateInfo,
  saveToken,
  closeDialog,
}) {
  const sendData = () => {
    axios
      .post("http://localhost:8080/api/auth/login", {
        ...info,
      })
      .then((response) => {
        console.log(response);
        successNotification("You are logged in!");
        saveToken(response.data.access_token);
        closeDialog();
      })
      .catch((error) => {
        console.log(error);
        if (error.response) {
          let str = error.response.data.message;
          errorNotification(str.charAt(0).toUpperCase() + str.slice(1) + ".");
        } else {
          errorNotification(error.message);
        }
      });
  };

  const onChange = (e) => {
    updateInfo(e.target.name, e.target.value.trim());
    validateInfo(e.target.name, e.target.value.trim());
  };
  const onSubmit = (e) => {
    e.preventDefault();
    console.log(info);

    if (!info["username"] || !info["password"]) {
      errorNotification("Please enter your username and password.");
      validateInfo("username", info["username"]);
      validateInfo("password", info["password"]);
    } else {
      sendData();
    }
  };

  return (
    <div className="App">
      <form
        noValidate
        autoComplete="off"
        onSubmit={(e) => onSubmit(e)}
        className="login-form"
      >
        <TextField
          className="text-field"
          onChange={onChange}
          name="username"
          label="Username"
          variant="outlined"
          size="small"
          error={!!errors["username"]}
          helperText={errors["username"]}
        />

        <TextField
          type="password"
          className="text-field"
          onChange={onChange}
          name="password"
          label="Password"
          variant="outlined"
          size="small"
          error={!!errors["password"]}
          helperText={errors["password"]}
        />

        <Button
          type="submit"
          fullWidth
          variant="contained"
          color="primary"
          className="form-button"
        >
          Sign In
        </Button>
      </form>
    </div>
  );
}

const mapStateToProps = (state) => ({
  info: state.login.login_info,
  errors: state.login.login_errors,
});

const mapDispatchToProps = (dispatch) => ({
  updateInfo: (type, data) => dispatch(actions.updateLoginInfo(type, data)),
  validateInfo: (type, data) => dispatch(actions.validateLoginInfo(type, data)),
  saveToken: (token) => dispatch(login(token)),
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginForm);
