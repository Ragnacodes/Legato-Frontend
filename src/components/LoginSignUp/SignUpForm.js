import React from "react";
import { useState, useEffect, useRef } from "react";
import axios from "axios";
import { connect } from "react-redux";
import * as actions from "../../actions/signup";
import url from '../../utils/api-url';

import {
  TextField,
  Button,
  Backdrop,
  CircularProgress,
  InputAdornment,
  IconButton,
} from "@material-ui/core";
import Visibility from "@material-ui/icons/Visibility";
import VisibilityOff from "@material-ui/icons/VisibilityOff";

import { successNotification, errorNotification } from "../Notification";

export function SignUpForm({
  info,
  errors,
  updateInfo,
  validateInfo,
  closeDialog,
  resetForm,
}) {
  const signupForm = useRef(null);
  const [showPassword, setShowPassword] = useState(false);
  const [showPasswordConfirm, setShowPasswordConfirm] = useState(false);
  const [loading, setLoading] = useState(false);
  const [send, setSend] = useState(false);

  const noError = () => {
    for (const [, value] of Object.entries(errors)) {
      if (value) return false;
    }
    return true;
  };

  useEffect(() => {
    if (send) {
      if (noError()) {
        sendData();
        console.log("send");
      } else {
        errorNotification("PLease enter valid data.");
      }
    }
    return () => {
      setSend(false);
    };
  });

  const sendData = () => {
    setLoading(true);
    axios
      .post(`${url()}/auth/signup`, {
        ...info,
        confirm: "",
      })
      .then((response) => {
        setLoading(false);
        closeDialog();
        console.log("response", response);
        signupForm.current.reset();
        resetForm();
        successNotification("You have created your account. Please log in.");
      })
      .catch((error) => {
        setLoading(false);
        if (error.response) {
          let str = error.response.data.message;
          errorNotification(str.charAt(0).toUpperCase() + str.slice(1) + ".");
        } else {
          errorNotification(error.message);
        }
        console.log("error", error.message);
      });
  };

  const validateAll = () => {
    validateInfo("email", info["email"]);
    validateInfo("password", info["password"]);
    validateInfo("confirm", info["confirm"]);
    validateInfo("username", info["username"]);
  };

  const onChange = (e) => {
    updateInfo(e.target.name, e.target.value.trim());
    validateInfo(e.target.name, e.target.value);
  };

  const onSubmit = (e) => {
    e.preventDefault();
    setSend(true);
    validateAll();
  };

  return (
    <div>
      <Backdrop className="signup-backdrop" open={loading}>
        {loading && <CircularProgress color="secondary" />}
      </Backdrop>

      <form
        ref={signupForm}
        noValidate
        autoComplete="off"
        onSubmit={(e) => onSubmit(e)}
        className="signup-form"
      >
        <TextField
          className="text-field"
          required
          onChange={onChange}
          name="username"
          error={!!errors["username"]}
          label="Username"
          helperText={errors["username"]}
          variant="outlined"
          size="small"
        />

        <TextField
          className="text-field"
          required
          onChange={onChange}
          name="email"
          error={!!errors["email"]}
          label="Email"
          helperText={errors["email"]}
          variant="outlined"
          size="small"
        />

        <TextField
          type={showPassword ? "text" : "password"}
          autoComplete="false"
          className="text-field"
          required
          onChange={onChange}
          name="password"
          error={!!errors["password"]}
          label="Password"
          helperText={errors["password"]}
          variant="outlined"
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <IconButton
                  aria-label="toggle password visibility"
                  onClick={() => setShowPassword((prev) => !prev)}
                  onMouseDown={(e) => e.preventDefault()}
                  edge="end"
                >
                  {showPassword ? <Visibility /> : <VisibilityOff />}
                </IconButton>
              </InputAdornment>
            ),
          }}
          size="small"
        />

        <TextField
          type={showPasswordConfirm ? "text" : "password"}
          className="text-field"
          required
          onChange={onChange}
          name="confirm"
          error={!!errors["confirm"]}
          label="Confirm Password"
          helperText={errors["confirm"]}
          variant="outlined"
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <IconButton
                  aria-label="toggle password visibility"
                  onClick={() => setShowPasswordConfirm((prev) => !prev)}
                  onMouseDown={(e) => e.preventDefault()}
                  edge="end"
                >
                  {showPasswordConfirm ? <Visibility /> : <VisibilityOff />}
                </IconButton>
              </InputAdornment>
            ),
          }}
          size="small"
        />

        <Button
          disabled={loading}
          type="submit"
          fullWidth
          variant="contained"
          className="form-button"
          color="primary"
        >
          Sign Up
        </Button>
      </form>
    </div>
  );
}

const mapStateToProps = (state) => ({
  info: state.signup.signup_info,
  errors: state.signup.signup_errors,
});

const mapDispatchToProps = (dispatch) => ({
  updateInfo: (type, data) => dispatch(actions.updateInfo(type, data)),
  validateInfo: (type, data) => dispatch(actions.validateInfo(type, data)),
  resetForm: () => dispatch(actions.resetSignupForm()),
});

export default connect(mapStateToProps, mapDispatchToProps)(SignUpForm);
