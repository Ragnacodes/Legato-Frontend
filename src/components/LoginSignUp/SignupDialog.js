import React from 'react';
import AbstractDialog from './AbstractDialog';
import SignupForm from './SignUpForm';

const SignupDialog = ({ signupOpen, setSignupOpen }) => {
    return (
        <AbstractDialog 
            open={signupOpen}
            onClick={setSignupOpen}
            title={'Sign Up'}
            description={'Please fill this form:'}
            form={<SignupForm />}
        />
    );
};

export default SignupDialog;