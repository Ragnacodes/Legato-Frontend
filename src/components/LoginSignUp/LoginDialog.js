import React from 'react';
import AbstractDialog from './AbstractDialog';
import LoginForm from './LoginForm';

const LoginDialog = ({ loginOpen, setLoginOpen }) => {
    return (
        <AbstractDialog 
            open={loginOpen}
            onClick={setLoginOpen}
            title={'Sign In'}
            description={'Welcome Back!'}
            form={<LoginForm />}
        />
    );
};

export default LoginDialog;