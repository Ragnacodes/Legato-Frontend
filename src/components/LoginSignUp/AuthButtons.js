import React, { useState } from 'react';
import { Button } from '@material-ui/core';
import LoginDialog from './LoginDialog';
import SignupDialog from './SignupDialog';

const AppbarButtons = () => {
    const [loginOpen, setLoginOpen] = useState(false);
    const [signupOpen, setSignupOpen] = useState(false);

    return (
        <>
            <Button
                color="secondary"
                onClick={() => setLoginOpen(true)}
                className="login-button"
            >
                Sign In
            </Button>
            <Button
                onClick={() => setSignupOpen(true)}
                variant="contained"
                color="secondary"
                className="signup-button"
            >
                Sign Up
            </Button>

            <LoginDialog loginOpen={loginOpen} setLoginOpen={setLoginOpen} />
            <SignupDialog signupOpen={signupOpen} setSignupOpen={setSignupOpen} />
        </>
    );
};

export default AppbarButtons;