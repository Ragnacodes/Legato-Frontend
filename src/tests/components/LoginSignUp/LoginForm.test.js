import React from "react";
import { shallow } from "enzyme";
import axios from "../__mocks__/axios";
import { LoginForm } from "../../../components/LoginSignUp/LoginForm";
import { TextField, Button } from "@material-ui/core";

let wrapper, props, validateInfo, updateInfo, info, errors, saveToken;
beforeEach(() => {
  info = {
    username: "",
    password: "",
  };
  errors = {
    username: "",
    password: "",
  };

  validateInfo = jest.fn();
  updateInfo = jest.fn();
  saveToken = jest.fn();
  props = {
    info,
    errors,
    validateInfo,
    updateInfo,
    saveToken,
  };
  wrapper = shallow(<LoginForm {...props} />);
});

test("should render LoginForm correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should render Text Fields correctly", () => {
  expect(wrapper.find(TextField).length).toBe(2);
});

test("should render Button correctly", () => {
  expect(wrapper.find(Button).length).toBe(1);
});

test("should disable button (username and password empty)", () => {
  const button = wrapper.find(Button);
  expect(button).toBeDisabled;
});

test("should update data correctly(password)", () => {
  const password = "11111111";
  const field = wrapper.find(TextField).find({ name: "password" });
  field.simulate("change", {
    target: {
      name: field.prop("name"),
      value: password,
    },
  });
  expect(updateInfo).toHaveBeenCalledWith("password", password);
  expect(validateInfo).toHaveBeenCalledWith("password", password);
});

test("should update data correctly(username)", () => {
  const username = "username";
  const field = wrapper.find(TextField).find({ name: "username" });
  field.simulate("change", {
    target: {
      name: field.prop("name"),
      value: username,
    },
  });
  expect(updateInfo).toHaveBeenCalledWith("username", username);
  expect(validateInfo).toHaveBeenCalledWith("username", username);
});

test("should login", (done) => {
  const username = "i";
  const password = "iiiiiiii";
  info = {
    username: username,
    password: password,
  };
  axios.post.mockImplementationOnce(
    () =>
      Promise.resolve({
        data: {
          access_token: "something",
        },
      }) //promise
  );

  wrapper = shallow(<LoginForm {...props} info={info} />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {},
  });
  setTimeout(() => {
    expect(saveToken).toHaveBeenCalledTimes(1);
    done();
  }, 0);
});
