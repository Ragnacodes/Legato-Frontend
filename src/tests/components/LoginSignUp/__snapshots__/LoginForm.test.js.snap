// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render LoginForm correctly 1`] = `
<div
  className="App"
>
  <form
    autoComplete="off"
    className="login-form"
    noValidate={true}
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(TextField))
      className="text-field"
      error={false}
      helperText=""
      label="Username"
      name="username"
      onChange={[Function]}
      size="small"
      variant="outlined"
    />
    <WithStyles(ForwardRef(TextField))
      className="text-field"
      error={false}
      helperText=""
      label="Password"
      name="password"
      onChange={[Function]}
      size="small"
      type="password"
      variant="outlined"
    />
    <WithStyles(ForwardRef(Button))
      className="form-button"
      color="primary"
      disabled={true}
      fullWidth={true}
      type="submit"
      variant="contained"
    >
      Sign In
    </WithStyles(ForwardRef(Button))>
  </form>
</div>
`;
